### PR TITLE
refactor(cell): Refactor shared to register loading using ldmatrix.

### PR DESCRIPTION
**Pull request by haruhi55**
Refactor the implementation of loading a shared memory tile into registers.

1. `xToXLoader/XToXStorer`: These are the highest-level interfaces that expose only very logical concepts, such as Row-major or Column-major Tiles on a specific memory hierarchy.
1. `xToXLoaderImpl/XToXStorerImpl`: These implement tile transfer between memory hierarchies. They are specialized based on the specific instruction used and the source and/or destination layout.
1. `xxStoreBase/xxLoadBase`: These are thin wrappers for specific instructions to transfer a single `BaseTile`.

- [x] Move the instruction-related implementation into `copy_atom.hpp`.
- [x] Fix load.
- [x] Fix store.
- [x] Enhance implementations based on the updated RegTile definition.

### CODE REVIEW

File: include/cell/copy/copy_atom.hpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    /// @brief a thin wrapper for executing ldmatrix instruction to load a
+    ///        `16x16` tile to register.
+    template <const int kRows, const int kCols>
+    DEVICE void ldmatrix(const DType* src, DType* dst) {
+        uint32_t* reg = reinterpret_cast<uint32_t*>(dst);
+        uint32_t smem_addr =
+            static_cast<uint32_t>(__cvta_generic_to_shared(src));
+
+        for (int i = 0; i < kRows; ++i) {
```
Comment (KuangjuX) [inline include/cell/copy/copy_atom.hpp:56]: Why are the two nested for loops necessary in the `ldmatrix` implementation?

File: tests/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -15,11 +15,6 @@ if(WITH_TESTING)
       continue() # the unittest for gemm requires extra dependencies
     endif()
 
-    # FIXME(haruhi): recover these commented unittest after refactor.
```
Comment (KuangjuX) [inline tests/cpp/CMakeLists.txt:18]: Can we re-add it now?

File: include/cell/copy/copy_atom.hpp (hunk 1 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp:56]: You're right. These loops are incorrect. A stupid error.

File: tests/cpp/CMakeLists.txt (hunk 2 above)
Comment (haruhi55) [reply tests/cpp/CMakeLists.txt:18]: This PR add the `test_s2r_copy` and pass the unittest.

There are other two unittests being deleted, test a single wmma, and test_gemm will be added in the next two PRs.

File: include/cell/copy/copy_atom.hpp (hunk 1 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: > Why are the two nested for loops necessary in the `ldmatrix` implementation?

Fixed.

Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: > Why are the two nested for loops necessary in the `ldmatrix` implementation?

The previous loops are to support the use of `ldmatrix` for loading data types other than half-precision tiles, but I didn't add unittest for it. After careful consideration and verification of the tensor core's `wmma` input type, I added a constraint specifying that `ldmatrix` should only be used for half-precision floating-point numbers.

**Review by KuangjuX — APPROVED** [review]
LGTM.

**Pull request merged**